File: frontend/src/types/recommendation.ts
/**
 * Types mirroring backend/app/schemas/profile.py and
 * backend/app/schemas/recommendation.py field-for-field (snake_case kept
 * on purpose — these cross the wire as-is via lib/api.ts).
 */

export type CalismaTipi = "ofis" | "hibrit" | "uzaktan";

export interface UserProfile {
  budget_m2?: number | null;
  has_children: boolean;
  has_car: boolean;
  elderly_in_household: boolean;
  deprem_onceligi: number;
  saglik_onceligi: number;
  egitim_onceligi: number;
  ulasim_onceligi: number;
  sosyal_onceligi: number;
  calisma_tipi: CalismaTipi;
  calisma_ilcesi?: string | null;
  office_lat?: number | null;
  office_lon?: number | null;
  max_commute_minutes?: number | null;
  free_text?: string | null;
}

export interface ScoreBreakdown {
  deprem_guvenlik: number;
  saglik: number;
  egitim: number;
  ulasim: number;
  sosyal_yasam: number;
  yasam_kalitesi: number;
}

export type ScoreCriterion = keyof ScoreBreakdown;

export type OfisUlasimModu = "karayolu" | "raylı_sistem" | "vapur";

export interface RawData {
  hastane_count: number;
  okul_count: number;
  cami_count: number;
  toplanma_count: number;
  ofis_mesafesi_km: number | null;
  ofis_tahmini_sure_dk: number | null;
  ofis_ulasim_modu: OfisUlasimModu | null;
}

export interface NeighborhoodResult {
  mahalle_id: string;
  mahalle_adi: string;
  ilce: string;
  uygunluk_skoru: number;
  score_breakdown: ScoreBreakdown;
  raw: RawData;
  avg_m2_fiyat: number | null;
  geometry: Record<string, unknown>;
  ai_aciklama: string | null;
  disclaimer: string;
}

export interface RecommendationResponse {
  top5: NeighborhoodResult[];
  alternatifler: NeighborhoodResult[];
  applied_weights: Record<string, number>;
  weight_source: "ai" | "rule_based";
  total_considered: number;
}

File: backend/app/schemas/recommendation.py
"""
Recommendation response schemas.

Responsibilities
----------------
- NeighborhoodResult  : tek mahalle önerisi (skor kırılımı + geometry + AI açıklama)
- RecommendationResponse : top-5 + alternatifler + uygulanan ağırlıklar
"""
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class ScoreBreakdown(BaseModel):
    """Her kriter için 0-100 arası bireysel skor."""
    deprem_guvenlik: float
    saglik:          float
    egitim:          float
    ulasim:          float
    sosyal_yasam:    float
    yasam_kalitesi:  float


class RawData(BaseModel):
    """Ham sayım verileri (açıklanabilirlik için)."""
    hastane_count:   int = 0
    okul_count:      int = 0
    cami_count:      int = 0
    toplanma_count:  int = 0
    ofis_mesafesi_km: float | None = Field(
        None,
        description=(
            "Kullanıcının işaretlediği ofis konumuna mesafe (km). Bu mahalle "
            "OSRM aday listesindeyse gerçek karayolu mesafesi, değilse kuş "
            "uçuşu tahmindir. Sadece office_lat/office_lon gönderildiyse hesaplanır."
        ),
    )
    ofis_tahmini_sure_dk: float | None = Field(
        None,
        description=(
            "Tahmini işe gidiş süresi (dakika). Karayolu/raylı sistem/vapur "
            "seçeneklerinden hesaplanabilenlerin en hızlısıdır (bkz. "
            "ofis_ulasim_modu); OSRM verisi yoksa kuş uçuşu kaba tahmindir."
        ),
    )
    ofis_ulasim_modu: Optional[Literal["karayolu", "raylı_sistem", "vapur"]] = Field(
        None,
        description=(
            "ofis_tahmini_sure_dk hangi ulaşım biçimiyle hesaplandı. Sadece "
            "gerçek OSRM verisi mevcutsa doludur — kuş uçuşu tahminde None."
        ),
    )


class NeighborhoodResult(BaseModel):
    """Tek mahalle öneri objesi."""

    mahalle_id:      str
    mahalle_adi:     str
    ilce:            str
    uygunluk_skoru:  float = Field(..., description="0-100 kompozit uygunluk skoru")
    score_breakdown: ScoreBreakdown
    raw:             RawData
    avg_m2_fiyat:    Optional[int] = Field(None, description="Tahmini TL/m2 fiyatı")
    geometry:        Dict[str, Any] = Field(..., description="GeoJSON Polygon")
    ai_aciklama:     Optional[str] = Field(
        None,
        description="Türkçe AI açıklaması (Gemini). Hata durumunda None olabilir.",
    )
    disclaimer: str = Field(
        default=(
            "⚠️ Bu sonuçlar istatistiksel bölge verilerine dayanmaktadır. "
            "Bina bazında güvenlik garantisi verilmez; mühendislik raporu veya "
            "zemin etüdü yerine geçmez."
        ),
        description="Her mahalle sonucunda görünmesi zorunlu sorumluluk reddi metni.",
    )


class RecommendationResponse(BaseModel):
    """POST /api/v1/recommend uç noktasının tam yanıtı."""

    top5:             List[NeighborhoodResult]
    alternatifler:    List[NeighborhoodResult] = Field(
        default_factory=list,
        description="6-8. sıra öneriler (karşılaştırma için)",
    )
    applied_weights:  Dict[str, float] = Field(
        ...,
        description="Skorlamada kullanılan kriter ağırlıkları (şeffaflık)",
    )
    weight_source:    str = Field(
        default="rule_based",
        description="'ai' veya 'rule_based' — ağırlık nereden geldi?",
    )
    total_considered: int = Field(
        ...,
        description="Bütçe filtresi sonrası değerlendirilen mahalle sayısı",
    )
